File: packages/ios-app/Sources/Models/Skill.swift
import Foundation

// MARK: - Skill Source

/// Where a skill was loaded from
enum SkillSource: String, Codable {
    case global  // ~/.tron/skills/
    case project // .tron/skills/ (relative to project)
}

// MARK: - Skill Model

/// Skill information for listing
struct Skill: Identifiable, Codable, Equatable {
    /// Skill name (folder name, used as @reference)
    let name: String
    /// Short description (first non-header line of SKILL.md)
    let description: String
    /// Where the skill was loaded from
    let source: SkillSource
    /// Whether this skill auto-injects into every prompt (Rules)
    let autoInject: Bool
    /// Tags for categorization
    let tags: [String]?

    var id: String { name }

    // Coding keys for JSON decoding
    private enum CodingKeys: String, CodingKey {
        case name, description, source, autoInject, tags
    }
}

// MARK: - Skill Metadata (Full Details)

/// Full skill metadata including content
struct SkillMetadata: Identifiable, Codable, Equatable {
    /// Skill name (folder name, used as @reference)
    let name: String
    /// Short description (first non-header line of SKILL.md)
    let description: String
    /// Where the skill was loaded from
    let source: SkillSource
    /// Whether this skill auto-injects into every prompt (Rules)
    let autoInject: Bool
    /// Tags for categorization
    let tags: [String]?
    /// Full SKILL.md content (after frontmatter stripped)
    let content: String
    /// Absolute path to skill folder
    let path: String
    /// List of additional files in the skill folder
    let additionalFiles: [String]

    var id: String { name }

    /// Convert to basic Skill info
    var asSkill: Skill {
        Skill(
            name: name,
            description: description,
            source: source,
            autoInject: autoInject,
            tags: tags
        )
    }
}

// MARK: - RPC Response Types

/// Response from skill.list RPC call
struct SkillListResponse: Codable {
    let skills: [Skill]
    let totalCount: Int
    let autoInjectCount: Int
}

/// Response from skill.get RPC call
struct SkillGetResponse: Codable {
    let skill: SkillMetadata?
    let found: Bool
}

/// Response from skill.refresh RPC call
struct SkillRefreshResponse: Codable {
    let success: Bool
    let skillCount: Int
}

File: packages/agent/src/rpc/types/skill.ts
/**
 * @fileoverview Skill RPC Types
 *
 * Types for skill operations methods.
 */

// =============================================================================
// Skill Methods
// =============================================================================

/**
 * Skill info returned in list operations
 */
export interface RpcSkillInfo {
  /** Skill name (folder name, used as @reference) */
  name: string;
  /** Human-readable display name (from frontmatter, falls back to folder name) */
  displayName: string;
  /** Short description (from frontmatter or first non-header line of SKILL.md) */
  description: string;
  /** Where the skill was loaded from */
  source: 'global' | 'project';
  /** Whether this skill auto-injects into every prompt (Rules) */
  autoInject: boolean;
  /** Tags for categorization */
  tags?: string[];
}

/**
 * Full skill metadata with content
 */
export interface RpcSkillMetadata extends RpcSkillInfo {
  /** Full SKILL.md content (after frontmatter stripped) */
  content: string;
  /** Absolute path to skill folder */
  path: string;
  /** List of additional files in the skill folder */
  additionalFiles: string[];
}

/** List available skills */
export interface SkillListParams {
  /** Session ID to get working directory for project skills */
  sessionId?: string;
  /** Filter by source (global, project) */
  source?: 'global' | 'project';
  /** Filter for auto-inject skills only */
  autoInjectOnly?: boolean;
  /** Include full content in results */
  includeContent?: boolean;
}

export interface SkillListResult {
  /** List of skills (with or without content based on includeContent param) */
  skills: RpcSkillInfo[] | RpcSkillMetadata[];
  /** Total number of skills */
  totalCount: number;
  /** Number of auto-inject skills (Rules) */
  autoInjectCount: number;
}

/** Get a single skill by name */
export interface SkillGetParams {
  /** Session ID to get working directory for project skills */
  sessionId?: string;
  /** Skill name */
  name: string;
}

export interface SkillGetResult {
  /** Skill metadata with full content */
  skill: RpcSkillMetadata | null;
  /** Whether the skill was found */
  found: boolean;
}

/** Refresh skills cache */
export interface SkillRefreshParams {
  /** Session ID to get working directory for project skills */
  sessionId?: string;
}

export interface SkillRefreshResult {
  /** Whether the refresh was successful */
  success: boolean;
  /** Number of skills loaded after refresh */
  skillCount: number;
}

/** Remove a skill from session context */
export interface SkillRemoveParams {
  /** Session ID */
  sessionId: string;
  /** Name of the skill to remove */
  skillName: string;
}

export interface SkillRemoveResult {
  /** Whether the skill was successfully removed */
  success: boolean;
  /** Error message if removal failed */
  error?: string;
}
